### PR TITLE
links interface and gRPC-on-wire implementation

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
First in a series building toward the beta API.

### CODE REVIEW

**Comment by nathanielmanistaatgoogle** [comment]
Rebasing this on top of the merge of #2249 let me drop a TODO from the test, so that was nice. :-)

**Comment by nathanielmanistaatgoogle** [comment]
Travis-green!

File: src/python/src/grpc/framework/interfaces/links/test_utilities.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""State and behavior appropriate for use in tests."""
+
+import threading
+
+from grpc.framework.interfaces.links import links
+
+
+class Link(links.Link):
```
Comment (soltanmm) [inline src/python/src/grpc/framework/interfaces/links/test_utilities.py]: Could we give this a more descriptive name?

File: src/python/src/grpc/_links/_proto_scenarios.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+  def requests(self):
+    """Access the sequence of requests for this scenario.
+
+    Returns:
+      A sequence of request protocol buffers.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def response_for_request(self, request):
```
Comment (soltanmm) [inline src/python/src/grpc/_links/_proto_scenarios.py]: Do we care about handling half-duplex-like connections for tests wherein we expect some number of responses after some other number of requests? Would this interface handle that if we do?

Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_links/_proto_scenarios.py]: We do. This interface would not handle those; we'd have to add that in later.

(Note that for this commit, this file is a branch of src/python/src/grpc/_adapter/_proto_scenarios.py with negligible changes.)

File: src/python/src/grpc/framework/interfaces/links/test_utilities.py (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/framework/interfaces/links/test_utilities.py]: Renamed "RecordingLink".

**Comment by nathanielmanistaatgoogle** [comment]
Rebased resolve conflict in python_test.json but not otherwise changed except where noted; please re-review.

**Comment by nathanielmanistaatgoogle** [comment]
(gcov C++ appears broken at project head at the moment; travis is green otherwise)

**Comment by nathanielmanistaatgoogle** [comment]
This may have just uncovered a timing sensitivity in _later_test, but is Jenkins-green otherwise. See https://grpc-testing.appspot.com/job/gRPC_pull_requests/864/ for a test run with _later_test passing.

**Pull request merged**